File: backend/app/schemas/theming.py
"""
Theming & White-Label Schemas

Pydantic schemas for themes, branding, custom CSS, and white-label configuration.
"""

from typing import Any, Optional
from pydantic import BaseModel, Field
from enum import Enum
from datetime import datetime


# Enums

class ThemeMode(str, Enum):
    """Theme color modes"""
    LIGHT = "light"
    DARK = "dark"
    AUTO = "auto"  # Follow system preference


class ThemeScope(str, Enum):
    """Theme scope"""
    GLOBAL = "global"  # System-wide
    WORKSPACE = "workspace"  # Per workspace
    USER = "user"  # Per user
    EMBED = "embed"  # For embedded content


class ComponentType(str, Enum):
    """UI component types for styling"""
    BUTTON = "button"
    INPUT = "input"
    CARD = "card"
    TABLE = "table"
    CHART = "chart"
    SIDEBAR = "sidebar"
    HEADER = "header"
    MODAL = "modal"
    TOOLTIP = "tooltip"
    BADGE = "badge"
    ALERT = "alert"
    DROPDOWN = "dropdown"


class FontWeight(str, Enum):
    """Font weights"""
    THIN = "100"
    LIGHT = "300"
    REGULAR = "400"
    MEDIUM = "500"
    SEMIBOLD = "600"
    BOLD = "700"
    EXTRABOLD = "800"


# Color Models

class ColorPalette(BaseModel):
    """Color palette configuration"""
    # Primary colors
    primary: str = "#3B82F6"  # Blue
    primary_hover: str = "#2563EB"
    primary_foreground: str = "#FFFFFF"

    # Secondary colors
    secondary: str = "#6B7280"  # Gray
    secondary_hover: str = "#4B5563"
    secondary_foreground: str = "#FFFFFF"

    # Accent colors
    accent: str = "#8B5CF6"  # Purple
    accent_hover: str = "#7C3AED"
    accent_foreground: str = "#FFFFFF"

    # Background colors
    background: str = "#FFFFFF"
    background_secondary: str = "#F9FAFB"
    background_tertiary: str = "#F3F4F6"

    # Foreground/text colors
    foreground: str = "#111827"
    foreground_secondary: str = "#6B7280"
    foreground_muted: str = "#9CA3AF"

    # Border colors
    border: str = "#E5E7EB"
    border_focus: str = "#3B82F6"

    # Status colors
    success: str = "#10B981"
    success_foreground: str = "#FFFFFF"
    warning: str = "#F59E0B"
    warning_foreground: str = "#000000"
    error: str = "#EF4444"
    error_foreground: str = "#FFFFFF"
    info: str = "#3B82F6"
    info_foreground: str = "#FFFFFF"

    # Chart colors (for data visualization)
    chart_colors: list[str] = Field(default_factory=lambda: [
        "#3B82F6", "#10B981", "#F59E0B", "#EF4444", "#8B5CF6",
        "#EC4899", "#06B6D4", "#84CC16", "#F97316", "#6366F1"
    ])


class DarkPalette(ColorPalette):
    """Dark mode color palette"""
    primary: str = "#60A5FA"
    primary_hover: str = "#3B82F6"

    background: str = "#111827"
    background_secondary: str = "#1F2937"
    background_tertiary: str = "#374151"

    foreground: str = "#F9FAFB"
    foreground_secondary: str = "#D1D5DB"
    foreground_muted: str = "#9CA3AF"

    border: str = "#374151"


# Typography Models

class FontConfig(BaseModel):
    """Font configuration"""
    family: str = "Inter, system-ui, sans-serif"
    fallback: str = "system-ui, -apple-system, sans-serif"
    weight: FontWeight = FontWeight.REGULAR
    size_base: str = "16px"
    line_height: str = "1.5"


class Typography(BaseModel):
    """Typography configuration"""
    # Font families
    font_sans: str = "Inter, system-ui, sans-serif"
    font_mono: str = "JetBrains Mono, monospace"
    font_heading: Optional[str] = None  # Uses sans by default

    # Font sizes
    size_xs: str = "0.75rem"
    size_sm: str = "0.875rem"
    size_base: str = "1rem"
    size_lg: str = "1.125rem"
    size_xl: str = "1.25rem"
    size_2xl: str = "1.5rem"
    size_3xl: str = "1.875rem"
    size_4xl: str = "2.25rem"

    # Line heights
    line_tight: str = "1.25"
    line_normal: str = "1.5"
    line_relaxed: str = "1.75"

    # Letter spacing
    tracking_tight: str = "-0.025em"
    tracking_normal: str = "0"
    tracking_wide: str = "0.025em"

    # Font weights
    weight_light: str = "300"
    weight_normal: str = "400"
    weight_medium: str = "500"
    weight_semibold: str = "600"
    weight_bold: str = "700"


# Spacing & Layout Models

class Spacing(BaseModel):
    """Spacing configuration"""
    base: str = "4px"
    xs: str = "4px"
    sm: str = "8px"
    md: str = "16px"
    lg: str = "24px"
    xl: str = "32px"
    xxl: str = "48px"


class BorderRadius(BaseModel):
    """Border radius configuration"""
    none: str = "0"
    sm: str = "4px"
    md: str = "8px"
    lg: str = "12px"
    xl: str = "16px"
    full: str = "9999px"


class Shadow(BaseModel):
    """Shadow configuration"""
    sm: str = "0 1px 2px rgba(0, 0, 0, 0.05)"
    md: str = "0 4px 6px rgba(0, 0, 0, 0.1)"
    lg: str = "0 10px 15px rgba(0, 0, 0, 0.1)"
    xl: str = "0 20px 25px rgba(0, 0, 0, 0.15)"


# Component Styles

class ComponentStyle(BaseModel):
    """Style overrides for a component"""
    component: ComponentType
    styles: dict[str, str] = Field(default_factory=dict)
    variants: dict[str, dict[str, str]] = Field(default_factory=dict)


# Theme Models

class Theme(BaseModel):
    """Complete theme configuration"""
    id: str
    name: str
    description: Optional[str] = None
    mode: ThemeMode = ThemeMode.LIGHT
    scope: ThemeScope = ThemeScope.GLOBAL

    # Colors
    colors: ColorPalette = Field(default_factory=ColorPalette)
    dark_colors: Optional[ColorPalette] = None  # For auto mode

    # Typography
    typography: Typography = Field(default_factory=Typography)

    # Layout
    spacing: Spacing = Field(default_factory=Spacing)
    border_radius: BorderRadius = Field(default_factory=BorderRadius)
    shadows: Shadow = Field(default_factory=Shadow)

    # Component overrides
    components: list[ComponentStyle] = Field(default_factory=list)

    # Custom CSS
    custom_css: Optional[str] = None

    # Metadata
    is_default: bool = False
    is_system: bool = False
    created_by: Optional[str] = None
    workspace_id: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: Optional[datetime] = None

    # Preview
    preview_image: Optional[str] = None


class ThemeCreate(BaseModel):
    """Create theme request"""
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    mode: ThemeMode = ThemeMode.LIGHT
    scope: ThemeScope = ThemeScope.WORKSPACE
    colors: Optional[ColorPalette] = None
    dark_colors: Optional[ColorPalette] = None
    typography: Optional[Typography] = None
    spacing: Optional[Spacing] = None
    border_radius: Optional[BorderRadius] = None
    shadows: Optional[Shadow] = None
    components: list[ComponentStyle] = Field(default_factory=list)
    custom_css: Optional[str] = None
    workspace_id: Optional[str] = None


class ThemeUpdate(BaseModel):
    """Update theme request"""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = None
    mode: Optional[ThemeMode] = None
    colors: Optional[ColorPalette] = None
    dark_colors: Optional[ColorPalette] = None
    typography: Optional[Typography] = None
    spacing: Optional[Spacing] = None
    border_radius: Optional[BorderRadius] = None
    shadows: Optional[Shadow] = None
    components: Optional[list[ComponentStyle]] = None
    custom_css: Optional[str] = None
    is_default: Optional[bool] = None


# Branding Models

class Logo(BaseModel):
    """Logo configuration"""
    light_url: Optional[str] = None  # Logo for light mode
    dark_url: Optional[str] = None  # Logo for dark mode
    favicon_url: Optional[str] = None
    width: int = 120
    height: int = 40


class Branding(BaseModel):
    """White-label branding configuration"""
    id: str
    workspace_id: str

    # Basic info
    company_name: str
    tagline: Optional[str] = None
    support_email: Optional[str] = None
    support_url: Optional[str] = None

    # Logos
    logo: Logo = Field(default_factory=Logo)

    # Colors (quick branding)
    primary_color: str = "#3B82F6"
    secondary_color: str = "#6B7280"

    # Custom domain
    custom_domain: Optional[str] = None
    custom_domain_verified: bool = False

    # Footer
    footer_text: Optional[str] = None
    footer_links: list[dict[str, str]] = Field(default_factory=list)

    # Legal
    privacy_policy_url: Optional[str] = None
    terms_of_service_url: Optional[str] = None

    # Social
    social_links: dict[str, str] = Field(default_factory=dict)

    # Email templates
    email_from_name: Optional[str] = None
    email_from_address: Optional[str] = None

    # Feature flags
    hide_powered_by: bool = False
    custom_login_page: bool = False
    custom_error_pages: bool = False

    # Active theme
    theme_id: Optional[str] = None

    # Timestamps
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: Optional[datetime] = None


class BrandingCreate(BaseModel):
    """Create branding configuration"""
    workspace_id: str
    company_name: str
    tagline: Optional[str] = None
    primary_color: Optional[str] = None
    secondary_color: Optional[str] = None
    logo: Optional[Logo] = None


class BrandingUpdate(BaseModel):
    """Update branding configuration"""
    company_name: Optional[str] = None
    tagline: Optional[str] = None
    support_email: Optional[str] = None
    support_url: Optional[str] = None
    logo: Optional[Logo] = None
    primary_color: Optional[str] = None
    secondary_color: Optional[str] = None
    custom_domain: Optional[str] = None
    footer_text: Optional[str] = None
    footer_links: Optional[list[dict[str, str]]] = None
    privacy_policy_url: Optional[str] = None
    terms_of_service_url: Optional[str] = None
    social_links: Optional[dict[str, str]] = None
    email_from_name: Optional[str] = None
    email_from_address: Optional[str] = None
    hide_powered_by: Optional[bool] = None
    theme_id: Optional[str] = None


# User Preferences

class UserThemePreferences(BaseModel):
    """User theme preferences"""
    user_id: str
    theme_mode: ThemeMode = ThemeMode.AUTO
    preferred_theme_id: Optional[str] = None
    font_size: str = "medium"  # small, medium, large
    reduce_motion: bool = False
    high_contrast: bool = False
    custom_overrides: dict[str, str] = Field(default_factory=dict)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


# CSS Generation

class CSSVariables(BaseModel):
    """CSS custom properties"""
    variables: dict[str, str]
    scope: str = ":root"


class GeneratedCSS(BaseModel):
    """Generated CSS output"""
    css: str
    variables: CSSVariables
    minified: bool = False
    hash: str  # For cache busting


# Response Models

class ThemeListResponse(BaseModel):
    """List of themes"""
    themes: list[Theme]
    total: int


class ThemePreviewResponse(BaseModel):
    """Theme preview data"""
    theme: Theme
    css: GeneratedCSS
    preview_components: dict[str, str]  # Component previews


# Constants

DEFAULT_LIGHT_THEME = Theme(
    id="system-light",
    name="Light",
    description="Default light theme",
    mode=ThemeMode.LIGHT,
    is_default=True,
    is_system=True,
)

DEFAULT_DARK_THEME = Theme(
    id="system-dark",
    name="Dark",
    description="Default dark theme",
    mode=ThemeMode.DARK,
    colors=DarkPalette(),
    is_system=True,
)

PRESET_THEMES = {
    "ocean": {
        "name": "Ocean",
        "primary": "#0891B2",
        "secondary": "#0E7490",
        "accent": "#06B6D4",
    },
    "forest": {
        "name": "Forest",
        "primary": "#059669",
        "secondary": "#047857",
        "accent": "#10B981",
    },
    "sunset": {
        "name": "Sunset",
        "primary": "#EA580C",
        "secondary": "#C2410C",
        "accent": "#F97316",
    },
    "berry": {
        "name": "Berry",
        "primary": "#BE185D",
        "secondary": "#9D174D",
        "accent": "#EC4899",
    },
    "slate": {
        "name": "Slate",
        "primary": "#475569",
        "secondary": "#334155",
        "accent": "#64748B",
    },
}


# Helper Functions

def generate_css_variables(colors: ColorPalette) -> dict[str, str]:
    """Generate CSS custom properties from color palette."""
    return {
        "--color-primary": colors.primary,
        "--color-primary-hover": colors.primary_hover,
        "--color-primary-foreground": colors.primary_foreground,
        "--color-secondary": colors.secondary,
        "--color-secondary-hover": colors.secondary_hover,
        "--color-secondary-foreground": colors.secondary_foreground,
        "--color-accent": colors.accent,
        "--color-accent-hover": colors.accent_hover,
        "--color-accent-foreground": colors.accent_foreground,
        "--color-background": colors.background,
        "--color-background-secondary": colors.background_secondary,
        "--color-background-tertiary": colors.background_tertiary,
        "--color-foreground": colors.foreground,
        "--color-foreground-secondary": colors.foreground_secondary,
        "--color-foreground-muted": colors.foreground_muted,
        "--color-border": colors.border,
        "--color-border-focus": colors.border_focus,
        "--color-success": colors.success,
        "--color-warning": colors.warning,
        "--color-error": colors.error,
        "--color-info": colors.info,
    }


def generate_typography_variables(typography: Typography) -> dict[str, str]:
    """Generate CSS variables from typography config."""
    return {
        "--font-sans": typography.font_sans,
        "--font-mono": typography.font_mono,
        "--font-heading": typography.font_heading or typography.font_sans,
        "--text-xs": typography.size_xs,
        "--text-sm": typography.size_sm,
        "--text-base": typography.size_base,
        "--text-lg": typography.size_lg,
        "--text-xl": typography.size_xl,
        "--text-2xl": typography.size_2xl,
        "--text-3xl": typography.size_3xl,
        "--text-4xl": typography.size_4xl,
    }


def validate_hex_color(color: str) -> bool:
    """Validate hex color format."""
    import re
    return bool(re.match(r'^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$', color))


def lighten_color(hex_color: str, percent: float) -> str:
    """Lighten a hex color by a percentage."""
    # Simple implementation - production would be more sophisticated
    hex_color = hex_color.lstrip('#')
    rgb = tuple(int(hex_color[i:i+2], 16) for i in (0, 2, 4))
    new_rgb = tuple(min(255, int(c + (255 - c) * percent)) for c in rgb)
    return '#{:02x}{:02x}{:02x}'.format(*new_rgb)


def darken_color(hex_color: str, percent: float) -> str:
    """Darken a hex color by a percentage."""
    hex_color = hex_color.lstrip('#')
    rgb = tuple(int(hex_color[i:i+2], 16) for i in (0, 2, 4))
    new_rgb = tuple(max(0, int(c * (1 - percent))) for c in rgb)
    return '#{:02x}{:02x}{:02x}'.format(*new_rgb)

File: frontend/src/types/theming.ts
/**
 * Theming & White-Label Types
 *
 * TypeScript types for themes, branding, CSS generation, and white-label configuration.
 */

// Enums

export enum ThemeMode {
  LIGHT = 'light',
  DARK = 'dark',
  AUTO = 'auto',
}

export enum ThemeScope {
  GLOBAL = 'global',
  WORKSPACE = 'workspace',
  USER = 'user',
  EMBED = 'embed',
}

export enum ComponentType {
  BUTTON = 'button',
  INPUT = 'input',
  CARD = 'card',
  TABLE = 'table',
  CHART = 'chart',
  SIDEBAR = 'sidebar',
  HEADER = 'header',
  MODAL = 'modal',
  TOOLTIP = 'tooltip',
  BADGE = 'badge',
  ALERT = 'alert',
  DROPDOWN = 'dropdown',
}

export enum FontWeight {
  THIN = '100',
  LIGHT = '300',
  REGULAR = '400',
  MEDIUM = '500',
  SEMIBOLD = '600',
  BOLD = '700',
  EXTRABOLD = '800',
}

// Color Models

export interface ColorPalette {
  // Primary colors
  primary: string;
  primary_hover: string;
  primary_foreground: string;

  // Secondary colors
  secondary: string;
  secondary_hover: string;
  secondary_foreground: string;

  // Accent colors
  accent: string;
  accent_hover: string;
  accent_foreground: string;

  // Background colors
  background: string;
  background_secondary: string;
  background_tertiary: string;

  // Foreground/text colors
  foreground: string;
  foreground_secondary: string;
  foreground_muted: string;

  // Border colors
  border: string;
  border_focus: string;

  // Status colors
  success: string;
  success_foreground: string;
  warning: string;
  warning_foreground: string;
  error: string;
  error_foreground: string;
  info: string;
  info_foreground: string;

  // Chart colors
  chart_colors: string[];
}

// Typography Models

export interface Typography {
  font_sans: string;
  font_mono: string;
  font_heading?: string | null;

  size_xs: string;
  size_sm: string;
  size_base: string;
  size_lg: string;
  size_xl: string;
  size_2xl: string;
  size_3xl: string;
  size_4xl: string;

  line_tight: string;
  line_normal: string;
  line_relaxed: string;

  tracking_tight: string;
  tracking_normal: string;
  tracking_wide: string;

  weight_light: string;
  weight_normal: string;
  weight_medium: string;
  weight_semibold: string;
  weight_bold: string;
}

// Layout Models

export interface Spacing {
  base: string;
  xs: string;
  sm: string;
  md: string;
  lg: string;
  xl: string;
  xxl: string;
}

export interface BorderRadius {
  none: string;
  sm: string;
  md: string;
  lg: string;
  xl: string;
  full: string;
}

export interface Shadow {
  sm: string;
  md: string;
  lg: string;
  xl: string;
}

// Component Style

export interface ComponentStyle {
  component: ComponentType;
  styles: Record<string, string>;
  variants: Record<string, Record<string, string>>;
}

// Theme Models

export interface Theme {
  id: string;
  name: string;
  description?: string | null;
  mode: ThemeMode;
  scope: ThemeScope;

  colors: ColorPalette;
  dark_colors?: ColorPalette | null;

  typography: Typography;
  spacing: Spacing;
  border_radius: BorderRadius;
  shadows: Shadow;

  components: ComponentStyle[];
  custom_css?: string | null;

  is_default: boolean;
  is_system: boolean;
  created_by?: string | null;
  workspace_id?: string | null;
  created_at: string;
  updated_at?: string | null;

  preview_image?: string | null;
}

export interface ThemeCreate {
  name: string;
  description?: string | null;
  mode?: ThemeMode;
  scope?: ThemeScope;
  colors?: Partial<ColorPalette>;
  dark_colors?: Partial<ColorPalette>;
  typography?: Partial<Typography>;
  spacing?: Partial<Spacing>;
  border_radius?: Partial<BorderRadius>;
  shadows?: Partial<Shadow>;
  components?: ComponentStyle[];
  custom_css?: string;
  workspace_id?: string;
}

export interface ThemeUpdate {
  name?: string;
  description?: string | null;
  mode?: ThemeMode;
  colors?: Partial<ColorPalette>;
  dark_colors?: Partial<ColorPalette>;
  typography?: Partial<Typography>;
  spacing?: Partial<Spacing>;
  border_radius?: Partial<BorderRadius>;
  shadows?: Partial<Shadow>;
  components?: ComponentStyle[];
  custom_css?: string;
  is_default?: boolean;
}

// Branding Models

export interface Logo {
  light_url?: string | null;
  dark_url?: string | null;
  favicon_url?: string | null;
  width: number;
  height: number;
}

export interface Branding {
  id: string;
  workspace_id: string;

  company_name: string;
  tagline?: string | null;
  support_email?: string | null;
  support_url?: string | null;

  logo: Logo;

  primary_color: string;
  secondary_color: string;

  custom_domain?: string | null;
  custom_domain_verified: boolean;

  footer_text?: string | null;
  footer_links: Array<Record<string, string>>;

  privacy_policy_url?: string | null;
  terms_of_service_url?: string | null;

  social_links: Record<string, string>;

  email_from_name?: string | null;
  email_from_address?: string | null;

  hide_powered_by: boolean;
  custom_login_page: boolean;
  custom_error_pages: boolean;

  theme_id?: string | null;

  created_at: string;
  updated_at?: string | null;
}

export interface BrandingCreate {
  workspace_id: string;
  company_name: string;
  tagline?: string;
  primary_color?: string;
  secondary_color?: string;
  logo?: Logo;
}

export interface BrandingUpdate {
  company_name?: string;
  tagline?: string;
  support_email?: string;
  support_url?: string;
  logo?: Logo;
  primary_color?: string;
  secondary_color?: string;
  custom_domain?: string;
  footer_text?: string;
  footer_links?: Array<Record<string, string>>;
  privacy_policy_url?: string;
  terms_of_service_url?: string;
  social_links?: Record<string, string>;
  email_from_name?: string;
  email_from_address?: string;
  hide_powered_by?: boolean;
  theme_id?: string;
}

// User Preferences

export interface UserThemePreferences {
  user_id: string;
  theme_mode: ThemeMode;
  preferred_theme_id?: string | null;
  font_size: string;
  reduce_motion: boolean;
  high_contrast: boolean;
  custom_overrides: Record<string, string>;
  updated_at: string;
}

// CSS Generation

export interface CSSVariables {
  variables: Record<string, string>;
  scope: string;
}

export interface GeneratedCSS {
  css: string;
  variables: CSSVariables;
  minified: boolean;
  hash: string;
}

// Response Models

export interface ThemeListResponse {
  themes: Theme[];
  total: number;
}

export interface ThemePreviewResponse {
  theme: Theme;
  css: GeneratedCSS;
  preview_components: Record<string, string>;
}

export interface PresetTheme {
  id: string;
  name: string;
  primary: string;
  secondary: string;
  accent: string;
}

// Constants

export const DEFAULT_COLORS: ColorPalette = {
  primary: '#3B82F6',
  primary_hover: '#2563EB',
  primary_foreground: '#FFFFFF',
  secondary: '#6B7280',
  secondary_hover: '#4B5563',
  secondary_foreground: '#FFFFFF',
  accent: '#8B5CF6',
  accent_hover: '#7C3AED',
  accent_foreground: '#FFFFFF',
  background: '#FFFFFF',
  background_secondary: '#F9FAFB',
  background_tertiary: '#F3F4F6',
  foreground: '#111827',
  foreground_secondary: '#6B7280',
  foreground_muted: '#9CA3AF',
  border: '#E5E7EB',
  border_focus: '#3B82F6',
  success: '#10B981',
  success_foreground: '#FFFFFF',
  warning: '#F59E0B',
  warning_foreground: '#000000',
  error: '#EF4444',
  error_foreground: '#FFFFFF',
  info: '#3B82F6',
  info_foreground: '#FFFFFF',
  chart_colors: [
    '#3B82F6', '#10B981', '#F59E0B', '#EF4444', '#8B5CF6',
    '#EC4899', '#06B6D4', '#84CC16', '#F97316', '#6366F1',
  ],
};

export const DARK_COLORS: ColorPalette = {
  ...DEFAULT_COLORS,
  primary: '#60A5FA',
  primary_hover: '#3B82F6',
  background: '#111827',
  background_secondary: '#1F2937',
  background_tertiary: '#374151',
  foreground: '#F9FAFB',
  foreground_secondary: '#D1D5DB',
  foreground_muted: '#9CA3AF',
  border: '#374151',
};

export const THEME_MODE_LABELS: Record<ThemeMode, string> = {
  [ThemeMode.LIGHT]: 'Light',
  [ThemeMode.DARK]: 'Dark',
  [ThemeMode.AUTO]: 'Auto (System)',
};

export const THEME_SCOPE_LABELS: Record<ThemeScope, string> = {
  [ThemeScope.GLOBAL]: 'Global',
  [ThemeScope.WORKSPACE]: 'Workspace',
  [ThemeScope.USER]: 'User',
  [ThemeScope.EMBED]: 'Embedded',
};

export const COMPONENT_TYPE_LABELS: Record<ComponentType, string> = {
  [ComponentType.BUTTON]: 'Button',
  [ComponentType.INPUT]: 'Input',
  [ComponentType.CARD]: 'Card',
  [ComponentType.TABLE]: 'Table',
  [ComponentType.CHART]: 'Chart',
  [ComponentType.SIDEBAR]: 'Sidebar',
  [ComponentType.HEADER]: 'Header',
  [ComponentType.MODAL]: 'Modal',
  [ComponentType.TOOLTIP]: 'Tooltip',
  [ComponentType.BADGE]: 'Badge',
  [ComponentType.ALERT]: 'Alert',
  [ComponentType.DROPDOWN]: 'Dropdown',
};

export const FONT_SIZE_OPTIONS = [
  { value: 'small', label: 'Small' },
  { value: 'medium', label: 'Medium' },
  { value: 'large', label: 'Large' },
];

export const PRESET_THEMES: Record<string, PresetTheme> = {
  ocean: {
    id: 'ocean',
    name: 'Ocean',
    primary: '#0891B2',
    secondary: '#0E7490',
    accent: '#06B6D4',
  },
  forest: {
    id: 'forest',
    name: 'Forest',
    primary: '#059669',
    secondary: '#047857',
    accent: '#10B981',
  },
  sunset: {
    id: 'sunset',
    name: 'Sunset',
    primary: '#EA580C',
    secondary: '#C2410C',
    accent: '#F97316',
  },
  berry: {
    id: 'berry',
    name: 'Berry',
    primary: '#BE185D',
    secondary: '#9D174D',
    accent: '#EC4899',
  },
  slate: {
    id: 'slate',
    name: 'Slate',
    primary: '#475569',
    secondary: '#334155',
    accent: '#64748B',
  },
};

// Helper Functions

export function isLightMode(mode: ThemeMode): boolean {
  if (mode === ThemeMode.LIGHT) return true;
  if (mode === ThemeMode.DARK) return false;
  // AUTO mode - check system preference
  return !window.matchMedia?.('(prefers-color-scheme: dark)').matches;
}

export function getEffectiveColors(theme: Theme): ColorPalette {
  if (theme.mode === ThemeMode.DARK && theme.dark_colors) {
    return theme.dark_colors;
  }
  if (theme.mode === ThemeMode.AUTO) {
    const prefersDark = window.matchMedia?.('(prefers-color-scheme: dark)').matches;
    if (prefersDark && theme.dark_colors) {
      return theme.dark_colors;
    }
  }
  return theme.colors;
}

export function validateHexColor(color: string): boolean {
  return /^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/.test(color);
}

export function lightenColor(hexColor: string, percent: number): string {
  const hex = hexColor.replace('#', '');
  const r = parseInt(hex.substring(0, 2), 16);
  const g = parseInt(hex.substring(2, 4), 16);
  const b = parseInt(hex.substring(4, 6), 16);

  const newR = Math.min(255, Math.round(r + (255 - r) * percent));
  const newG = Math.min(255, Math.round(g + (255 - g) * percent));
  const newB = Math.min(255, Math.round(b + (255 - b) * percent));

  return `#${newR.toString(16).padStart(2, '0')}${newG.toString(16).padStart(2, '0')}${newB.toString(16).padStart(2, '0')}`;
}

export function darkenColor(hexColor: string, percent: number): string {
  const hex = hexColor.replace('#', '');
  const r = parseInt(hex.substring(0, 2), 16);
  const g = parseInt(hex.substring(2, 4), 16);
  const b = parseInt(hex.substring(4, 6), 16);

  const newR = Math.max(0, Math.round(r * (1 - percent)));
  const newG = Math.max(0, Math.round(g * (1 - percent)));
  const newB = Math.max(0, Math.round(b * (1 - percent)));

  return `#${newR.toString(16).padStart(2, '0')}${newG.toString(16).padStart(2, '0')}${newB.toString(16).padStart(2, '0')}`;
}

export function hexToRgb(hex: string): { r: number; g: number; b: number } | null {
  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  return result
    ? {
        r: parseInt(result[1], 16),
        g: parseInt(result[2], 16),
        b: parseInt(result[3], 16),
      }
    : null;
}

export function rgbToHex(r: number, g: number, b: number): string {
  return `#${[r, g, b].map((x) => x.toString(16).padStart(2, '0')).join('')}`;
}

export function getContrastColor(hexColor: string): string {
  const rgb = hexToRgb(hexColor);
  if (!rgb) return '#000000';

  // Calculate relative luminance
  const luminance = (0.299 * rgb.r + 0.587 * rgb.g + 0.114 * rgb.b) / 255;
  return luminance > 0.5 ? '#000000' : '#FFFFFF';
}

export function generateColorPalette(primaryColor: string): Partial<ColorPalette> {
  return {
    primary: primaryColor,
    primary_hover: darkenColor(primaryColor, 0.1),
    primary_foreground: getContrastColor(primaryColor),
    accent: lightenColor(primaryColor, 0.2),
    accent_hover: primaryColor,
    accent_foreground: getContrastColor(lightenColor(primaryColor, 0.2)),
    border_focus: primaryColor,
    info: primaryColor,
    info_foreground: getContrastColor(primaryColor),
  };
}

export function cssVariablesToObject(cssText: string): Record<string, string> {
  const variables: Record<string, string> = {};
  const regex = /--([a-zA-Z0-9-]+):\s*([^;]+);/g;
  let match;

  while ((match = regex.exec(cssText)) !== null) {
    variables[`--${match[1]}`] = match[2].trim();
  }

  return variables;
}

export function objectToCSSVariables(
  variables: Record<string, string>,
  scope = ':root'
): string {
  const entries = Object.entries(variables)
    .map(([key, value]) => `  ${key}: ${value};`)
    .join('\n');

  return `${scope} {\n${entries}\n}`;
}

// Theme State Management

export interface ThemingState {
  themes: Theme[];
  activeTheme: Theme | null;
  presets: PresetTheme[];
  branding: Branding | null;
  userPreferences: UserThemePreferences | null;
  isLoading: boolean;
  error: string | null;
  selectedThemeId: string | null;
  previewMode: boolean;
  editingTheme: Theme | null;
}

export function createInitialThemingState(): ThemingState {
  return {
    themes: [],
    activeTheme: null,
    presets: Object.values(PRESET_THEMES),
    branding: null,
    userPreferences: null,
    isLoading: false,
    error: null,
    selectedThemeId: null,
    previewMode: false,
    editingTheme: null,
  };
}

// CSS Application Helpers

export function applyThemeToDocument(theme: Theme): void {
  const colors = getEffectiveColors(theme);
  const root = document.documentElement;

  // Apply color variables
  Object.entries(colors).forEach(([key, value]) => {
    if (key !== 'chart_colors' && typeof value === 'string') {
      root.style.setProperty(`--color-${key.replace(/_/g, '-')}`, value);
    }
  });

  // Apply chart colors
  if (colors.chart_colors) {
    colors.chart_colors.forEach((color, index) => {
      root.style.setProperty(`--chart-color-${index + 1}`, color);
    });
  }

  // Apply typography
  const { typography } = theme;
  root.style.setProperty('--font-sans', typography.font_sans);
  root.style.setProperty('--font-mono', typography.font_mono);
  if (typography.font_heading) {
    root.style.setProperty('--font-heading', typography.font_heading);
  }

  // Apply spacing
  const { spacing } = theme;
  Object.entries(spacing).forEach(([key, value]) => {
    root.style.setProperty(`--spacing-${key}`, value);
  });

  // Apply border radius
  const { border_radius } = theme;
  Object.entries(border_radius).forEach(([key, value]) => {
    root.style.setProperty(`--radius-${key}`, value);
  });

  // Apply shadows
  const { shadows } = theme;
  Object.entries(shadows).forEach(([key, value]) => {
    root.style.setProperty(`--shadow-${key}`, value);
  });

  // Apply custom CSS if present
  if (theme.custom_css) {
    let customStyleEl = document.getElementById('theme-custom-css');
    if (!customStyleEl) {
      customStyleEl = document.createElement('style');
      customStyleEl.id = 'theme-custom-css';
      document.head.appendChild(customStyleEl);
    }
    customStyleEl.textContent = theme.custom_css;
  }

  // Set theme mode attribute
  root.setAttribute('data-theme', theme.mode);
  if (theme.mode === ThemeMode.DARK) {
    root.classList.add('dark');
  } else if (theme.mode === ThemeMode.LIGHT) {
    root.classList.remove('dark');
  } else {
    // AUTO mode
    const prefersDark = window.matchMedia?.('(prefers-color-scheme: dark)').matches;
    root.classList.toggle('dark', prefersDark);
  }
}

export function removeThemeFromDocument(): void {
  const root = document.documentElement;

  // Remove custom style element
  const customStyleEl = document.getElementById('theme-custom-css');
  if (customStyleEl) {
    customStyleEl.remove();
  }

  // Reset theme attribute
  root.removeAttribute('data-theme');
}
